### PR TITLE
added capability for running (e)BOSS data (issue #4)

### PR DESCRIPTION
**Pull request by ngbusca**
I've written the module boss.py (similar to desi.py) to input (e)boss data. The calling signature differs from that of rrdesi:

`
rrboss --spplate <spPlate> --spall <spAll> [--use-frames]
`

It runs over one single spPlate at a time. spAll is required to find the target ids (thing_ids) from plate-fiberid. 

The resolution matrix is built from wdisp assuming a gaussian shape and ndiag = int(4*wdisp_max + 1) ("4-sigma").

If --use-frames is not set, the spectra are read from the spPlate (this corresponds to the "coadd" option in rrdesi).

If --use-frames is set, the code will find the individual exposures from the spPlate header and get the spectra from the spCFrame files, assumed to be in the same directory as the spPlate. Unlike for desi, the wavelength grid is different for each spectrum. Redrock will see this a having multiple bands so it doesn't pose a problem per se. However, when the target class tries to coadd the individual exposures it will simply recopy all the input data. **this needs to be addresses before merging**

The implementation of parallelism is refactored in this branch: each parallel job receives a fraction of the list of targets and scans over all redshifts while in master each job receives all targes and scans over a fraction of the redshifts. This change addresses issue #35. 

Objects with THING_ID == -1 (typically sky but also some ELG, ..., which were selected with different photometry) are given a TARGETID = PLATEMJDFIBERID. Do we want to reject sky fibers and/or calibration stars?

### CODE REVIEW

**Comment by ngbusca** [comment]
I haven't tested with mpi (I don't know how to)

**Comment by ngbusca** [comment]
I've added a flag to the Target class to avoid coadding. When the flag is set, the target class won't get the .coadd attribute.

**Comment by sbailey** [comment]
This looks good except for the how THING_ID vs. (PLATE,MJD,FIBER) is mapped to TARGETID depending upon whether THING_ID is -1 or not.  Could we always use `targetid = plate*1000000000 + mjd*10000 + fiber` and ignore THING_ID (which I think would also mean we no longer need spAll either)?  It would be best to provide these conversions as two functions for end users to swap back and forth (double check these!)
```python
def platemjdfiber2targetid(plate, mjd, fiber):
    return plate*1000000000 + mjd*10000 + fiber

def targetid2platemjdfiber(targetid):
    fiber = targetid % 10000
    mjd = (targetid // 10000) % 100000
    plate = (targetid // (10000 * 100000))
    return (plate, mjd, fiber)
```

For rrdesi, I tested the parallelize over targets vs. redshifts at NERSC.  For small inputs (30 targets) the new form was 50% slower, but for large inputs (600 targets) they were within 3%.  That's fine.  At NERSC I don't see the memory difference that originally motivated this swap, but I'm willing to accept that the multiprocessing shared memory tricks work differently (worse) on other machines than at NERSC and on my laptop.

**Comment by ngbusca** [comment]
@sbailey sure, I'll change the THING_ID as you propose. What tool do you use to check the memory footprint? Maybe I'm being fooled by top.

**Comment by sbailey** [comment]
I based my memory estimate on the "maximum resident set size" reported by `/usr/bin/time rrdesi ...`.  Various tools have ambiguity in how they report shared memory usage by multiple processes, e.g.,
* (total memory) / (number of processes) so that the sum equals total memory used, but each process is reported as using less than it actually needs if it was running alone.
* (total memory used by the process) so that each process is reported as the amount that it needs if it was running alone, but has double counting across processes.

I haven't been able to find which one top is reporting.  And it is also possible that /usr/bin/time isn't counting the memory of spawned processes at all, in which case it wouldn't really be reporting what we want.

Bottom line: if you were having memory problems and this fixes it, ok, that's great.

**Comment by ngbusca** [comment]
Refactored the TARGETID as proposed above. There are still two open questions:

* There's no special treatment for SKY fibers (i.e. they are passed to zscan as physical spectra). This might be useful as a null test. Should we instead remove SKY fibers?
* I added an option to the Target class to avoid the coaddition step. This option is used by rrboss to skip this step, since it results in duplication of the data set (for (e)BOSS data).

**Comment by sbailey** [comment]
No special treatment for sky fibers is the same as what BOSS does, and is indeed useful as a null test.

Target option to avoid coaddition is fine.

Thanks for standardizing on (plate,mjd,fiber) <-> targetid mapping for rrboss.

Merging now.
